### PR TITLE
Threshold optimizer v1

### PR DESCRIPTION
**Pull request by mawuliadjei**
Wrapped all metrics in a class with documentation

### CODE REVIEW

**Review by yumanfai — CHANGES_REQUESTED** [review]
Overall I think we can go ahead but let's fix the parameter names so that we don't need to refactor after we publish

Other comments that we can work on in next iterations:
- search space using distinct y_score
- verbose using logging
- combine individual get_best_* functions into a optimize_metrics function

File: budgies/threshold_optimizer.py [hunk 1]
```diff
@@ -1,3 +1,251 @@
...
+    def set_search_space(self,
+                         search_space_size: int):
+        """set the number of possible probability threshold values to optimze for
+
+        This function is useful to reset the size of the search space after initializing the ThresholdOptimizer object.
+
+        Args:
+            search_space_size: The number of possible probability threshold values to optimze for
+        """
+        self.search_space = np.linspace(0, 1, search_space_size)
```
[the remark this answers is not in the document]
Comment (mawuliadjei) [reply budgies/threshold_optimizer.py:43]: Hmmm. I'm inclined to disagree. It is good thinking and I agree with the thought behind what you're saying. But:
1. It's better only when the dataset is small. When the dataset is large, you'll be looping values that would be of size N where N is the number of observations to predict from. In the core thought, the O notation of this looping process is O(N) which is linear, hence does not matter much.
2. Using the distinct values of the probabilities would mean that in the case where this method is applied on a validation set before implementing on test data, the values of the probabilities in the validation predictions, are going to skew the possible threshold values. I'm not exactly sure about the cost of this, but the risk makes sense to me.
3. Using the distinct values of the y_score does not give the user any control over the number of values to search over.

I think a good compromise will be to do something like:
`min_threshold = min(y_score)
 max_threshold = max(y_score)
self.search_space = np.linspace(min_threshold, max_threshold, search_space_size)`

This solution should meet the needs of points 1 and 3. And I guess 2 is kind of a moot point.

File: budgies/threshold_optimizer.py [hunk 2]
```diff
@@ -1,3 +1,251 @@
...
+
 import pandas as pd
 import numpy as np
-from sklearn.metrics import accuracy_score, f1_score
+from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, precision_score, recall_score
+
+
+class ThresholdOptimizer:
+    def __init__(self,
+                 predicted_probabilities: Union[np.ndarray, pd.Series, list],
```
[the remark this answers is not in the document]
Comment (mawuliadjei) [reply budgies/threshold_optimizer.py:10]: Agree

File: budgies/threshold_optimizer.py [hunk 3]
```diff
@@ -1,3 +1,251 @@
...
+                    'best_threshold': best_threshold,
+                    'all_scores': scores,
+                },
+            },
+        )
+        if verbose:
+            print(f'best {metric_type}: {best_score} occurs at threshold {best_threshold}')
+        return best_score, best_threshold
+
+    def get_best_f1_metrics(self,
```
[the remark this answers is not in the document]
Comment (mawuliadjei) [reply budgies/threshold_optimizer.py:97]: Agree. I kept it that way because we didn't ever get the time to discuss what we each did. I built it in such a way that anyone can take look at the function they worked on and maintain a high-level of readability. I'm aligned with you though.

File: budgies/threshold_optimizer.py [hunk 4]
```diff
@@ -1,3 +1,251 @@
...
+        self.optimized_metrics.update(
+            {
+                metric_type: {
+                    'best_score': best_score,
+                    'best_threshold': best_threshold,
+                    'all_scores': scores,
+                },
+            },
+        )
+        if verbose:
```
[the remark this answers is not in the document]
Comment (mawuliadjei) [reply budgies/threshold_optimizer.py:93]: yup. Aligned with you on this.

File: budgies/threshold_optimizer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yumanfai) [reply budgies/threshold_optimizer.py:43]: - Not sure if I understand point 2. Why does it skew the possible threshold values? Talking about handling for validation set, I think that'll be a topic later when we consider compatibility with things like Cross Validation.
- Another common situation for real life datasets is that y_score concentrate at certain ranges (e.g. between 0.4-0.6 / 0.7-0.9), with only some outliers on the far ends. In such cases, evenly distributed search space won't be helpful.

Looks like a common efficiency vs accuracy problem here. Good discussion and I think we can start with your suggested code first.
Ultimately we might consider to have a parameter to allow the user specifying which one is the priority, accuracy vs efficiency

[the remark this answers is not in the document]
Comment (mawuliadjei) [reply budgies/threshold_optimizer.py:43]: Point 2 is exactly what you're saying with the real life datasets. But on second thought, I think I'm over thinking it. So we move on. I'll commit the suggested changes and continue.
